bigtable/metrics.go

```diff
@@ -0,0 +1,426 @@
+		metric.WithExplicitBucketBoundaries(bucketBounds...),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Create server_latencies
+	tf.serverLatencies, err = meter.Float64Histogram(
+		metricNameServerLatencies,
+		metric.WithDescription("The latency measured from the moment that the RPC entered the Google data center until the RPC was completed."),
+		metric.WithUnit("ms"),
+		metric.WithExplicitBucketBoundaries(bucketBounds...),
+	)
+	if err != nil {
+		return err
+	}
+
+	// Create retry_count
+	tf.retryCount, err = meter.Int64Counter(
+		metricNameRetryCount,
+		metric.WithDescription("The number of additional RPCs sent after the initial attempt."),
+	)
+	return err
+}
+
+// builtinMetricsTracer is created one per operation
+// It is used to store metric instruments, attribute values
+// and other data required to obtain and record them
+type builtinMetricsTracer struct {
+	ctx            context.Context
+	builtInEnabled bool
+
+	// attributes that are specific to a client instance and
+	// do not change across different operations on client
+	clientAttributes []attribute.KeyValue
+
+	instrumentOperationLatencies metric.Float64Histogram
+	instrumentServerLatencies    metric.Float64Histogram
+	instrumentAttemptLatencies   metric.Float64Histogram
+	instrumentRetryCount         metric.Int64Counter
+
+	tableName   string
+	method      string
+	isStreaming bool
+
+	currOp opTracer
+}
+
+type opTracer struct {
+	attemptCount int64
+
+	startTime time.Time
+
+	// gRPC status code
+	status string
+
+	currAttempt attemptTracer
+}
+
+type attemptTracer struct {
+	startTime time.Time
+
+	// Contains the header response metadata which is used to extract cluster, zone and server latency
+	headerMD *metadata.MD
+
+	// Contains the trailer response metadata which is used to extract cluster, zone and server latency
+	trailerMD *metadata.MD
+
+	// gRPC status code
+	status string
+}
+
+func (tf *builtinMetricsTracerFactory) newBuiltinMetricsTracer(ctx context.Context, tableName string, isStreaming bool) builtinMetricsTracer {
+	return builtinMetricsTracer{
+		ctx:            ctx,
+		builtInEnabled: tf.builtinEnabled,
+
+		// Operation has started but not the attempt.
+		// So, create only operation tracer and not attempt tracer
+		currOp: opTracer{
+			startTime: time.Now(),
+		},
+		clientAttributes: tf.clientAttributes,
+
+		instrumentOperationLatencies: tf.operationLatencies,
+		instrumentServerLatencies:    tf.serverLatencies,
+		instrumentAttemptLatencies:   tf.attemptLatencies,
+		instrumentRetryCount:         tf.retryCount,
+
+		tableName:   tableName,
+		isStreaming: isStreaming,
+	}
+}
+
+// recordOperationCompletion records as many metrics as it can and does not return error
+func recordOperationCompletion(mt *builtinMetricsTracer) {
+	if !mt.builtInEnabled {
+		return
+	}
+
+	// Use headers and trailers from last attempt to extract cluster and zone
+	headerMD := mt.currOp.currAttempt.headerMD
+	trailerMD := mt.currOp.currAttempt.trailerMD
```

Comment:
I think it would be cleaner for the client push the parsed values instead of the tracer trying to parse them

Reply:
Done